src/content/insertable-streams/webgpu/js/multi_video_main.js

```diff
@@ -0,0 +1,274 @@
+            mappedAtCreation: true,
+        });
+        // Copies rectVerts to vertexBuffer
+        new Float32Array(vertexBuffer.getMappedRange()).set(rectVerts);
+        vertexBuffer.unmap();
+        this.vertexBuffer_ = vertexBuffer;
+
+        context.configure({
+            device,
+            format: swapChainFormat
+        })
+
+        this.renderPipeline_ = device.createRenderPipeline({
+            vertex: {
+                module: device.createShaderModule({
+                    code: wgslShaders.vertex,
+                }),
+                entryPoint: 'main',
+                buffers: [
+                    {
+                        arrayStride: 20,
+                        attributes: [
+                            {
+                                // position
+                                shaderLocation: 0,
+                                offset: 0,
+                                format: 'float32x3',
+                            },
+                            {
+                                // uv
+                                shaderLocation: 1,
+                                offset: 12,
+                                format: 'float32x2',
+                            },
+                        ],
+                    },
+                ],
+            },
+            fragment: {
+                module: device.createShaderModule({
+                    code: wgslShaders.fragment,
+                }),
+                entryPoint: 'main',
+                targets: [
+                    {
+                        format: swapChainFormat,
+                    },
+                ],
+            },
+            primitive: {
+                topology: 'triangle-list',
+            },
+        });
+
+        this.videoTexture_ = device.createTexture({
+            size: {
+                width: 480 * 2,
+                height: 270 * 2,
+                depthOrArrayLayers: 1,
+            },
+            arrayLayerCount: 1,
+            mipLevelCount: 1,
+            sampleCount: 1,
+            dimension: '2d',
+            format: 'rgba8unorm',
+            usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.TEXTURE_BINDING |
+                GPUTextureUsage.RENDER_ATTACHMENT,
+        });
+
+        this.sampler_ = device.createSampler({
+            addressModeU: 'repeat',
+            addressModeV: 'repeat',
+            addressModeW: 'repeat',
+            magFilter: 'linear',
+            minFilter: 'linear',
+        });
+    }
+
+    async renderOnScreen(videoFrame, gumFrame) {
+        const device = this.device_;
+        if (device === null || device === undefined) {
+            console.log('[WebGPUTransform] device is undefined or null.')
+            if (videoFrame) videoFrame.close();
+            if (gumFrame) gumFrame.close();
+            return false;
+        }
+        const videoTexture = this.videoTexture_;
+        let videoBitmap, gumBitmap;
+        if (videoFrame) {
```

Comment:
note: In the future we should be able to use GPUExternalTexture to avoid extra copies through an ImageBitmap. See the gPaste (though it requires a WIP Chromium CL)